### PR TITLE
Emailing workflow

### PR DESCRIPTION
**Pull request by paulinembabu**
See Issue https://github.com/OpenSRP/opensrp-client-emmunize-malawi/issues/8

### CODE REVIEW

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/FragmentBaseActivity.java [hunk 1]
```diff
@@ -73,13 +174,28 @@ protected void onCreate(Bundle savedInstanceState) {
...
+            if (report_date != null) {
+                DateFormat df = new SimpleDateFormat("dd MMM yyyy");
+                Date date = null;
+                try {
+                    date = df.parse(report_date);
+                } catch (ParseException e) {
+                    e.printStackTrace();
+                }
+                emailSubject = String.format("%s - %s - %s", name, new SimpleDateFormat("dd/MM/yyyy hh:mm:ss").format(date), communityName);
+                emailAttachmentName = String.format("%s-%s-%s", name, new SimpleDateFormat("ddMMyyyy hh:mm:ss").format(date), communityName);
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/FragmentBaseActivity.java:193]: Should these two have the same value?

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/util/PdfGeneratorHelper.java [hunk 2]
```diff
@@ -0,0 +1,553 @@
...
+            LayoutInflater inflater = (LayoutInflater)
+                    context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
+
+            for (int layout : layoutList) {
+                View generatedView = inflater.inflate(layout, null);
+                viewList.add(generatedView);
+            }
+
+        } catch (Exception e) {
+            Timber.e(e.getMessage());
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/util/PdfGeneratorHelper.java:74]: You can leave it as `Timber.e(e)` ... Could you catch a less general exception?

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/FragmentBaseActivity.java [hunk 3]
```diff
@@ -44,6 +64,87 @@ public static void startMe(Activity activity, String fragmentName, String title,
...
+                    , GizConstants.RQ_CODE.STORAGE_PERMISIONS)) {
+                final String[] pdfPath = {Environment.getExternalStorageDirectory() + File.separator + emailAttachmentName + ".pdf"};
+                // export as image
+                PdfGeneratorHelper.getBuilder()
+                        .setContext(this)
+                        .fromViewSource()
+                        .fromView(getPDFRootView())
+                        .setDefaultPageSize(PdfGeneratorHelper.PageSize.WRAP_CONTENT)
+                        .setFileName(emailAttachmentName)
+                        .setFolderName("Reports")
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/FragmentBaseActivity.java:80]: you can move this constants

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/FragmentBaseActivity.java [hunk 4]
```diff
@@ -44,6 +64,87 @@ public static void startMe(Activity activity, String fragmentName, String title,
...
+                        .openPDFafterGeneration(false)
+                        .build(new PdfGeneratorListener() {
+                            @Override
+                            public void onFailure(PdfGeneratorHelper.FailureResponse failureResponse) {
+                                super.onFailure(failureResponse);
+                                failureResponse.getThrowable().printStackTrace();
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/FragmentBaseActivity.java:86]: You can use `Timber` here

**Review by bennsimon — CHANGES_REQUESTED** [review]

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/FragmentBaseActivity.java [hunk 5]
```diff
@@ -73,13 +174,28 @@ protected void onCreate(Bundle savedInstanceState) {
...
+            String name = getRequestedFragmentName(fragmentName);
+
+            //Format Subject Date
+            if (report_date != null) {
+                DateFormat df = new SimpleDateFormat("dd MMM yyyy");
+                Date date = null;
+                try {
+                    date = df.parse(report_date);
+                } catch (ParseException e) {
+                    e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (paulinembabu) [reply opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/FragmentBaseActivity.java:190]: Done

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/FragmentBaseActivity.java (hunk 1 above)
Comment (paulinembabu) [reply opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/FragmentBaseActivity.java:193]: No, The Requirement was to have them as above.

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/FragmentBaseActivity.java (hunk 3 above)
Comment (paulinembabu) [reply opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/FragmentBaseActivity.java:80]: Done

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/FragmentBaseActivity.java (hunk 4 above)
Comment (paulinembabu) [reply opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/FragmentBaseActivity.java:86]: Done

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/util/PdfGeneratorHelper.java [hunk 6]
```diff
@@ -71,7 +71,7 @@ public static ContextStep getBuilder() {
             }
 
         } catch (Exception e) {
-            Timber.e(e.getMessage());
+            Timber.e("Could not inflate Layout");
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/util/PdfGeneratorHelper.java:74]: Could this be `Timber.e(e, "Could not inflate Layout");`

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/FragmentBaseActivity.java [hunk 7]
```diff
@@ -186,8 +185,8 @@ protected void onCreate(Bundle savedInstanceState) {
                 Date date = null;
                 try {
                     date = df.parse(report_date);
-                } catch (ParseException e) {
-                    e.printStackTrace();
+                } catch (Exception e) {
+                    Timber.e("Date could not be formatted");
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/FragmentBaseActivity.java:189]: Could this be `Timber.e(e, "Date could not be formatted");`

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/FragmentBaseActivity.java [hunk 8]
```diff
@@ -77,13 +76,13 @@ public void sendEmail() {
...
                         .setFileName(emailAttachmentName)
-                        .setFolderName("Reports")
+                        .setFolderName(GizConstants.EmailParameterHelper.REPORTS)
                         .openPDFafterGeneration(false)
                         .build(new PdfGeneratorListener() {
                             @Override
                             public void onFailure(PdfGeneratorHelper.FailureResponse failureResponse) {
                                 super.onFailure(failureResponse);
-                                failureResponse.getThrowable().printStackTrace();
+                                Timber.e("Pdf Could not be Generated");
```
Comment (bennsimon) [inline opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/FragmentBaseActivity.java:85]: Could this be `Timber.e(failureResponse.getThrowable(), "Pdf Could not be Generated");`

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/FragmentBaseActivity.java (hunk 7 above)
Comment (bennsimon) [reply opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/FragmentBaseActivity.java:189]: Leave it as `ParseException`

**Review by bennsimon — CHANGES_REQUESTED** [review]

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/FragmentBaseActivity.java (hunk 7 above)
Comment (paulinembabu) [reply opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/FragmentBaseActivity.java]: DONE

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/FragmentBaseActivity.java (hunk 8 above)
Comment (paulinembabu) [reply opensrp-giz-malawi/src/main/java/org/smartregister/giz/activity/FragmentBaseActivity.java]: DONE

File: opensrp-giz-malawi/src/main/java/org/smartregister/giz/util/PdfGeneratorHelper.java (hunk 6 above)
Comment (paulinembabu) [reply opensrp-giz-malawi/src/main/java/org/smartregister/giz/util/PdfGeneratorHelper.java]: Done

**Review by bennsimon — APPROVED** [review]

**Pull request merged**